### PR TITLE
muxers/mplex: Allow up to 60 bit long stream IDs

### PR DESCRIPTION
**Pull request by whereistejas**
hi @mxinden,
I have opened this PR for closing issue: #2061

Do we also need write a test to check if we can actually use 60 bit stream ID?

### CODE REVIEW

**Comment by mxinden** [comment]
> Do we also need write a test to check if we can actually use 60 bit stream ID?

Yes please, that would be great!

**Comment by whereistejas** [comment]
Hi @mxinden,
I'm having a hard time figuring out how to write a test for this. I tried to go through the programs in the examples directory, looking for something that I could use. Can you please give me a little guidance about how I should proceed with writing this test?

**Comment by mxinden** [comment]
`Codec` implements both `Encode` and `Decode`. To test your patch I would encode a `Frame::Open` with a  stream ID `> 2^32` into a `BytesMut` and then decode said `BytesMut` expecting the two stream IDs to equal.

Does that help @whereistejas?

**Comment by whereistejas** [comment]
Hi @mxinden ,
I have finished writing the test. Your suggestion helped me a bunch, thanks! :)

File: muxers/mplex/src/codec.rs [hunk 1]
```diff
@@ -305,4 +305,33 @@ mod tests {
         let ok_msg = Frame::Data { stream_id: LocalStreamId { num: 123, role }, data };
         assert!(enc.encode(ok_msg, &mut out).is_ok());
     }
+
+    #[test]
+    fn test_60bit_stream_id() {
+        // create new codec object for encoding and decoding our frame
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:311]:
```suggestion
        // Create new codec object for encoding and decoding our frame.
```

File: muxers/mplex/src/codec.rs [hunk 2]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+        let mut codec = Codec::new();
+        // create a u64 stream ID
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:313]:
```suggestion
        // Create a u64 stream ID.
```

File: muxers/mplex/src/codec.rs [hunk 3]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+        let id : u64 = u32::MAX as u64 + 1 ;
+        let stream_id = LocalStreamId { num: id, role: Endpoint::Dialer };
+
+        // open a new frame with that stream ID
+        let original_frame = Frame::Open { stream_id };
+
+        // encode that frame
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:320]:
```suggestion
        // Encode that frame.
```

File: muxers/mplex/src/codec.rs [hunk 4]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+        let mut enc_frame = BytesMut::new();
+        match codec.encode(original_frame, &mut enc_frame) {
+            Err(e) => panic!("{}", e.to_string()),
+            _ => {}
+        };
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:325]:
```suggestion
        codec.encode(original_frame, &mut enc_frame).expect("Encoding to succeed.");
```

File: muxers/mplex/src/codec.rs [hunk 5]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+
+        // decode encoded frame and extract stream ID
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:327]:
```suggestion
        // Cecode encoded frame and extract stream ID.
```

File: muxers/mplex/src/codec.rs [hunk 6]
```diff
@@ -305,4 +305,33 @@ mod tests {
         let ok_msg = Frame::Data { stream_id: LocalStreamId { num: 123, role }, data };
         assert!(enc.encode(ok_msg, &mut out).is_ok());
     }
+
+    #[test]
+    fn test_60bit_stream_id() {
+        // create new codec object for encoding and decoding our frame
+        let mut codec = Codec::new();
+        // create a u64 stream ID
+        let id : u64 = u32::MAX as u64 + 1 ;
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:314]:
```suggestion
        let id: u64 = u32::MAX as u64 + 1 ;
```

File: muxers/mplex/src/codec.rs [hunk 7]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+        match codec.encode(original_frame, &mut enc_frame) {
+            Err(e) => panic!("{}", e.to_string()),
+            _ => {}
+        };
+
+        // decode encoded frame and extract stream ID
+        let dec_string_id = match codec.decode(&mut enc_frame) {
+            Err(e) => panic!("{}", e.to_string()),
+            Ok(decoded_frame) => decoded_frame.unwrap().remote_id(),
+        };
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:331]:
```suggestion
        let dec_string_id = codec.decode(&mut enc_frame)
            .expect("Decoding to succeed.")
            .map(|f| f.remote_id())
            .unwrap();
```

File: muxers/mplex/src/codec.rs [hunk 8]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+        
+        assert_eq!(dec_string_id.num, stream_id.num);
+
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:334]:
```suggestion
```

File: muxers/mplex/src/codec.rs [hunk 9]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+        // decode encoded frame and extract stream ID
+        let dec_string_id = match codec.decode(&mut enc_frame) {
+            Err(e) => panic!("{}", e.to_string()),
+            Ok(decoded_frame) => decoded_frame.unwrap().remote_id(),
+        };
+        
+        assert_eq!(dec_string_id.num, stream_id.num);
+        
+    }
+
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:336]:
```suggestion
```

File: muxers/mplex/src/codec.rs [hunk 10]
```diff
@@ -305,4 +305,33 @@ mod tests {
...
+    #[test]
+    fn test_60bit_stream_id() {
+        // create new codec object for encoding and decoding our frame
+        let mut codec = Codec::new();
+        // create a u64 stream ID
+        let id : u64 = u32::MAX as u64 + 1 ;
+        let stream_id = LocalStreamId { num: id, role: Endpoint::Dialer };
+
+        // open a new frame with that stream ID
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs:317]:
```suggestion
        // Open a new frame with that stream ID.
```

**Review by mxinden — COMMENTED** [review]
Test looks great. Just a couple of smaller suggestions.

Would you mind including a changelog entry in `muxers/mplex/CHANGELOG.md`?

**Comment by whereistejas** [comment]
> Would you mind including a changelog entry in `muxers/mplex/CHANGELOG.md`?
Are we currently on version 0.39.0? I'm asking so that I may add the version in the change logs.

File: muxers/mplex/CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,7 @@
+# 0.39.0 [unreleased]
+
+- Increased the lenght of Stream ID to 60 bits.
```
Comment (mxinden) [inline muxers/mplex/CHANGELOG.md:3]: Please move this line under the `# 0.29.0 [...]` section below the `- Update dependencies` line.

File: muxers/mplex/src/codec.rs [hunk 12]
```diff
@@ -18,14 +18,10 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use asynchronous_codec::{Decoder, Encoder};
```
Comment (mxinden) [inline muxers/mplex/src/codec.rs]: Please revert the code formatting introduced in https://github.com/libp2p/rust-libp2p/pull/2094/commits/a01622283045025a33f6578082b5ed461c7e0372.

I am not opposed to using rustfmt, though I think, in case we do decide to use it across the repository, it should be a separate pull request. See https://github.com/libp2p/rust-libp2p/issues/1668 for details.

Comment (whereistejas) [reply muxers/mplex/src/codec.rs]: I had accidentally used `rustfmt` in the commit [ead5d97](https://github.com/libp2p/rust-libp2p/pull/2094/commits/ead5d97ae7d18dbc86abe7f5087d8e5deba00608). The commit [a016222](https://github.com/libp2p/rust-libp2p/pull/2094/commits/a01622283045025a33f6578082b5ed461c7e0372) reverted the formatting that was done by `rustfmt`.

**Comment by whereistejas** [comment]
I think, we are pretty much finished. I would love to take up `rustfmt` issue in a seperate PR. The least I can do is open the PR. I will leave the decision of whether or not to merge it to the reviewers.

File: muxers/mplex/CHANGELOG.md [hunk 13]
```diff
@@ -1,6 +1,7 @@
 # 0.29.0 [unreleased]
 
 - Update dependencies.
+- Increased the length of Stream ID to 60 bits.
```
Comment (mxinden) [inline muxers/mplex/CHANGELOG.md:4]:
```suggestion

- Support stream IDs of up to 60 bit length. See [PR 2094] for details.

[PR 2094]: https://github.com/libp2p/rust-libp2p/pull/2094
```

**Review by mxinden — APPROVED** [review]
Thanks for the help @whereistejas.

**Comment by whereistejas** [comment]
Thanks @mxinden! :)

**Comment by senden9** [comment]
Hi!
I know that this pull request is already merged. But if I look at the changed files I see no restriction to 60 Bit. As far as I can see there is nothing that would stop me from creating a 61, 62, 63 or 64 Bit ID.

**Comment by whereistejas** [comment]
@senden9 
I can be completely wrong here, but my understanding is that the first 3 bits are reserved for the message flag. Please refer to this document: https://github.com/libp2p/specs/blob/master/mplex/README.md

@mxinden can you please help out here? Also, I think it makes sense to add a test, that fails if we try to use a stream ID greater than 60bits.

**Comment by senden9** [comment]
Right. It seams that the first three bits are reserved for flags. And the lowest three bits are shifted away in encoding. So `LocalStreamId::num` seams not to be the ID, but the Header (as called your linked document).

Anyway as I want introduce any more "noise" in the notification inboxes of the people of this PR: I will think a bit about that and then create a new Discussion or Issue if anything seems strange to me. Thanks for the fast reply @whereistejas

**Comment by mxinden** [comment]
> I know that this pull request is already merged.

Feedback is always welcome, no matter what status the pull request is in ;)

> But if I look at the changed files I see no restriction to 60 Bit. As far as I can see there is nothing that would stop me from creating a 61, 62, 63 or 64 Bit ID.

Correct. You can still call `RemoteStreamId::dialer(u64::MAX)`. I agree that this is not ideal and that there should be something stopping you, e.g. through an `assert`.

On the other hand:

- I doubt anyone will ever reach stream IDs of more than 60 bit [given that it is monotonically increased by 1](https://github.com/libp2p/rust-libp2p/blob/c1ef4bffd225a78cefe8fa43dc8ee18e03ff4f93/muxers/mplex/src/codec.rs#L93-L98).
- `LocalStreamId` is not exposed by the crate, and thus an internal construct only.


All that said, the goal of this pull request is to support stream IDs of up to 60 bit coming from a remote peer. Using a `u64` instead of a `u32` does allow that.

This is the relevant excerpt from the specification:

> The maximum header length is 9 bytes (per the unsigned-varint spec). With 9 continuation bits and 3 message flag bits the maximum stream ID is 60 bits (maximum value of 2^60 - 1).

https://github.com/libp2p/specs/blob/master/mplex/README.md#message-format


@senden9 does the above make sense?

**Comment by senden9** [comment]
@mxinden yes, thanks for the explanation. The "Internal construct only"-part is also a good argument for why it isn't too bad to not block the constructing or reaching of that invalid state.
